[WIP] Support delay allreduce in DDP

mrshenli opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#35083 [WIP] Support delay allreduce in DDP**
* #36054 Fix DDP error checking for unused parameters

This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

mrshenli committed:
[WIP] Support delay allreduce in DDP

This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Closes #24005

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Closes #24005

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Closes #24005

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Closes #24005

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Closes #24005

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

mrshenli committed:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -264,6 +280,53 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
+      std::lock_guard<std::mutex> lock(this->mutex_);
+
+      const auto replica_count = replicas_.size();
+      grad_accumulators_.resize(replica_count);
```
nit: it is not used?

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.cpp:
```diff
@@ -264,6 +280,53 @@ void Reducer::mark_variable_ready_sparse(VariableIndex index) {
+              .variable_index = variable_index,
+          };
+
+          mark_variable_ready(index);
```
I'm wondering what is the performance difference comparing to just directly call all reduce to all gradients here? The main benefit to do bucket all reduce here probably is that it can work with round robin process groups.

zhaojuanmao commented on torch/csrc/distributed/c10d/reducer.h:
```diff
@@ -74,6 +75,11 @@ class Reducer {
   std::vector<VariableIndex> unused_parameters_;
+
+  const bool delay_allreduce_;
+  bool require_final_hook_;
```
nit: require_delayed_grad_hook_?

zhaojuanmao reviewed (commented):
overall looks good! just some nits and questions

Commit:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

Commit:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

Commit:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

Commit:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

Commit:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]

Commit:
Update on "[WIP] Support delay allreduce in DDP"


This mode delays gradients allreduce till the end of the backward
pass, i.e., there will be no overlap between gradient communication
and local backward computation. This will be slower than the
default mode, but should have less constraints on input models.

Differential Revision: [D20559668](https://our.internmc.facebook.com/intern/diff/D20559668)

[ghstack-poisoned]
